src/v/http/client.cc

```diff
@@ -97,7 +97,28 @@ static ss::future<iobuf> fail_on_error(const boost::beast::error_code& ec) {
     return ss::make_exception_future<iobuf>(except);
 }
 
+ss::future<> client::response_stream::prefetch_headers() {
+    if (is_header_done()) {
+        return ss::now();
+    }
+    return ss::do_with(iobuf(), [this](iobuf& head) {
+        return ss::do_until(
+                 [this] { return is_header_done(); },
+                 [this, &head] {
+                     return recv_some().then(
+                       [&head](iobuf buf) { head.append(std::move(buf)); });
+                 })
+          .then([this, &head] { _prefetch = std::move(head); });
+    });
+}
+
 ss::future<iobuf> client::response_stream::recv_some() {
+    if (_prefetch) {
```

Comment:
i didn't know iobuf has `operator bool()` overloaded.

Reply:
It's optional<iobuf>, I'll change that to raw iobuf. No need for extra indirection layer.